pkg/client/client.go

```diff
@@ -761,6 +763,126 @@ func (c *Client) WaitForPrometheus(ctx context.Context, p *monv1.Prometheus) err
 	return nil
 }
 
+// ValidationError indicates if an operation performed by the client has
+// resulted in an invalid state.
+type ValidationError struct {
+	Degraded    bool
+	Unavailable bool
+	Reasons     []string
+}
+
+func (ve *ValidationError) Error() string {
+	return fmt.Sprintf("error: %s degraded: %t unavailable: %t",
+		strings.Join(ve.Reasons, ";"), ve.Degraded, ve.Unavailable)
+}
+
+var _ error = (*ValidationError)(nil)
+
+func (c *Client) validatePrometheusPodState(ctx context.Context, p *monv1.Prometheus) *ValidationError {
+
+	status, _, err := prometheusoperator.Status(ctx, c.kclient.(*kubernetes.Clientset), p)
+	if err != nil {
+		klog.V(4).ErrorS(err, "validatePrometheusPodState: failed to get Prometheus status")
+		// TODO(sthaha): do we want to not set Degraded in this case?
+		return &ValidationError{Degraded: true, Reasons: []string{"failed to get Prometheus status"}}
+	}
+
+	// NOTE: about status (which seems a bit confusing)
+	// Available       = pods in Running && Ready  **but** MAY NOT be up to date
+	// UpdatedReplicas = Pods are Running && Ready && matches Spec
+	// Unavailable		 = Pods that not Available; i.e. any state other than Running & Ready
+	//
+	// Total Pods = Available + Unavailable
+	// Pods Running but NEEDS updation = Available - UpdatedReplicas (also returned as []pods)
+	// NeedsUpdation = Available - UpdatedReplicas
+
+	required := *p.Spec.Replicas
+	available := status.AvailableReplicas
+	uptoDate := status.UpdatedReplicas
+
+	needsUpdate := available - uptoDate
+
+	if uptoDate >= required || needsUpdate >= required {
```

Comment:
Comparing `available` and `required` directly should yield the same result?
```suggestion
	if available >= required {
```

Reply:
+1 since these pods are already running